Multipart enhancement

JustinKyleJames opened the pull request:
This is the for multipart enhancement described in https://github.com/irods/irods_client_s3_api/issues/114.

I have also implemented AbortMultipartUpload.

Notes:

- Tests are passing.
- I had previously made all of the changes for PR #115.  I opened a new PR simply because so many changes have been made since that PR was opened.
- This is now working for both normal clients and clients that use chunked encoding.
- I made a couple of minor updates not directly related to this enhancement:
    - I changed __FUNCTION__ references to __func__. This was only for these two files that are touched in this PR.
    - I changed some logging from debug to trace.

A couple of other concerns/questions:

The entry in shared memory is cleaned up when CompleteMultipartUpload or AbortMultipartUpload returns. However, if neither is called it won't be cleaned up until the server restarts.  A new call to InitiateMultipartUpload would generate a new upload_id so that upload should behave correctly.

trel commented:
> The entry in shared memory is cleaned up when CompleteMultipartUpload or AbortMultipartUpload returns. However, if neither is called it won't be cleaned up until the server restarts. 

Should we consider cleaning up the entry in shared memory if a certain amount of time passes?   Aka a timeout on idle shared memory entries?  If so... what is a safe amount?  An hour of no activity?  A day?  A week?

trel commented on README.md:
````diff
@@ -496,6 +496,8 @@ docker compose build
 docker compose run client
 ```
 
+> Note that if you get an error like "'name' does not match any of the regexes: '^x-'" then you will need to upgrade your version of docker-compose.
````
i think the hyphenated use is now deprecated / old.

```suggestion
> Note that if you get an error like "'name' does not match any of the regexes: '^x-'" then you will need to upgrade your version of docker compose.
```

JustinKyleJames replied:
Done

trel commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -717,10 +1020,16 @@ void manually_parse_chunked_body_write_to_irods_in_background(
 				}
-				else {
+				if (!keep_dstream_open_flag && d->is_open()) {
+					logging::debug("{}:{} Closing iRODS file.", __func__, __LINE__);
```
i think this is correct...

```suggestion
					logging::debug("{}:{} Closing iRODS data object.", __func__, __LINE__);
```

JustinKyleJames replied:
Changed all reference from "iRODS file" to "iRODS data object" in all files.  I also changed all open/close messages to trace instead of debug.

trel commented on endpoints/s3/src/completemultipartupload.cpp:
```diff
@@ -404,27 +432,48 @@ void irods::s3::actions::handle_completemultipartupload(
+		auto transport_ptr = std::get<3>(part_shmem::replica_token_and_number_map[upload_id]);
+		auto dstream_ptr = std::get<4>(part_shmem::replica_token_and_number_map[upload_id]);
+
+		logging::debug("{}:{} Closing iRODS file.", __func__, __LINE__);
```
perhaps?

```suggestion
		logging::debug("{}:{} Closing iRODS data object.", __func__, __LINE__);
```

JustinKyleJames replied:
done

trel commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -58,64 +98,135 @@ void manually_parse_chunked_body_write_to_irods_in_background(
+			if (part_shmem::replica_token_and_number_map.find(upload_id_) ==
+			    part_shmem::replica_token_and_number_map.end()) {
+				logging::debug(
+					"{}: Open new iRODS file [{}] for writing and seeking to {}.", __func__, irods_path_, part_offset_);
```
data object?

```suggestion
					"{}: Open new iRODS data object [{}] for writing and seeking to {}.", __func__, irods_path_, part_offset_);
```

JustinKyleJames replied:
done

trel commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -58,64 +98,135 @@ void manually_parse_chunked_body_write_to_irods_in_background(
+			else {
+				// get the replica token and pass it to open
+				logging::debug(
+					"{}: Open iRODS file [{}] for writing and seeking to {}.  Replica token={}",
```
data object?

```suggestion
					"{}: Open iRODS data object [{}] for writing and seeking to {}.  Replica token={}",
```

JustinKyleJames replied:
done

trel commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -58,64 +98,135 @@ void manually_parse_chunked_body_write_to_irods_in_background(
-			if (!odstream_) {
-				auto msg = fmt::format("{}: Failed to open irods file for writing [{}].", __func__, _irods_path);
+		else { // just a single part upload, stream directly to iRODS
+			logging::debug("{}: Open iRODS file [{}] for writing.", __func__, part_filename_);
```
```suggestion
			logging::debug("{}: Open iRODS data object [{}] for writing.", __func__, part_filename_);
```

trel commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -198,19 +309,20 @@ class incremental_async_read : public std::enable_shared_from_this<incremental_a
 				}
-				logging::debug(
+				logging::trace(
 					"{}: multipart upload: Wrote [{}] bytes to iRODS file [{}].",
```
```suggestion
					"{}: multipart upload: Wrote [{}] bytes to iRODS data object [{}].",
```

JustinKyleJames replied:
done

trel commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -198,19 +309,20 @@ class incremental_async_read : public std::enable_shared_from_this<incremental_a
-				else {
-					self->odstream_.close();
+				if (self->odstream_->is_open() && !self->keep_dstream_open_flag) {
+					logging::debug("{}:{} Closing iRODS file [{}].", __func__, __LINE__, self->irods_path_);
```
```suggestion
					logging::debug("{}:{} Closing iRODS data object [{}].", __func__, __LINE__, self->irods_path_);
```

JustinKyleJames replied:
done

JustinKyleJames replied:
done

trel commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -402,33 +640,82 @@ void irods::s3::actions::handle_putobject(
+				if (part_shmem::replica_token_and_number_map.find(upload_id) ==
+				    part_shmem::replica_token_and_number_map.end()) {
+					logging::debug(
+						"{}: Open new iRODS file [{}] for writing and seeking to {}.",
```
```suggestion
						"{}: Open new iRODS data object [{}] for writing and seeking to {}.",
```

JustinKyleJames replied:
done

trel commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -402,33 +640,82 @@ void irods::s3::actions::handle_putobject(
+				else {
+					// get the replica token and pass it to open
+					logging::debug(
+						"{}: Open iRODS file [{}] for writing and seeking to {}.",
```
```suggestion
						"{}: Open iRODS data object [{}] for writing and seeking to {}.",
```

JustinKyleJames replied:
done

trel reviewed (commented):
very impressive.   glad this is behaving.

we have existing tests?  new tests?

alanking commented on endpoints/s3/src/completemultipartupload.cpp:
```diff
@@ -404,27 +432,48 @@ void irods::s3::actions::handle_completemultipartupload(
 
+	// remove the temporary part files - on failures we don't want to clean up as this could be resent
+	for (int i = 0; i < max_part_number; ++i) {
+		std::remove(part_info_vector[i].part_filename.c_str());
```
Is this missing some parameters? I don't see any `std::remove` overloads like this... https://en.cppreference.com/w/cpp/algorithm/remove

JustinKyleJames replied:
This is the std::remove() to remove a file from here:

https://en.cppreference.com/w/cpp/io/c/remove

JustinKyleJames replied:
Trying to resolve this.

alanking commented on tests/docker/docker-compose.yml:
```diff
@@ -1,4 +1,5 @@
 version: '3'
+name: s3-api
```
Maybe "irods-s3-api"?

JustinKyleJames replied:
I made that change.

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -18,18 +18,57 @@
+	// for the first stream that is opened. These shared pointers are saved to this map so that the first
+	// open() to an iRODS data object can remain open throughout the lifecycle of the request.  This
+	// stream will be closed in either CompleteMultipartUpload or AbortMultipartUpload.
+	std::unordered_map<
+		std::string,
+		std::tuple<
+			irods::experimental::io::replica_token,
+			irods::experimental::io::replica_number,
+			std::shared_ptr<irods::experimental::client_connection>,
+			std::shared_ptr<irods::experimental::io::client::native_transport>,
+			std::shared_ptr<irods::experimental::io::odstream>>>
```
Consider adding some type aliases for things like this and even the type for `part_size_map`. Would warrant a Great Naming...

JustinKyleJames replied:
I am going to hold off on the alias until we can discuss naming.

korydraughn replied:
Please open an issue for this and link it to this conversation.

We can address this in a later release.

JustinKyleJames replied:
Issue opened.

alanking replied:
https://github.com/irods/irods_client_s3_api/issues/123

korydraughn replied:
Issue number is #123.

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -58,64 +98,135 @@ void manually_parse_chunked_body_write_to_irods_in_background(
 	std::shared_ptr<beast::http::request_parser<boost::beast::http::buffer_body>> parser_;
 	std::string irods_path_;
 	bool upload_part_flag_;
+	bool know_part_offset_flag_;
```
Does this mean that we (a) *want to know* the part offset, we (b) *do know* the part offset, or that we (c) *want to know whether we know* the part offset? It looks like it's (b). Should this be called `part_offset_is_known_`?

JustinKyleJames replied:
It is b and will change to that.

JustinKyleJames replied:
done

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -58,64 +98,135 @@ void manually_parse_chunked_body_write_to_irods_in_background(
+	{
+		if (keep_dstream_open_flag) {
+			//std::this_thread::sleep_for(std::chrono::seconds(3));
+			logging::debug("Destructor iRODS called for object that did first open");
```
I think if this is just for debugging, it might be better to just remove the destructor. If it seems like it will be really useful, consider surrounding this with a try/catch(...) so that no exceptions escape. Here's what cppreference says (with my own emphasis)... https://en.cppreference.com/w/cpp/language/destructor

> Although [std::uncaught_exceptions](https://en.cppreference.com/w/cpp/error/uncaught_exception) may sometimes be used to detect stack unwinding in progress, **it is generally considered bad practice to allow *any* destructor to terminate by throwing an exception**.

JustinKyleJames replied:
I am going to remove it.  I didn't intend on keeping this in.

JustinKyleJames replied:
done

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -198,19 +309,20 @@ class incremental_async_read : public std::enable_shared_from_this<incremental_a
 					self->session_ptr_->send(std::move(self->resp_)); // Schedules an async write op.
 					return;
 				}
-				logging::debug(
+				logging::trace(
 					"{}: multipart upload: Wrote [{}] bytes to iRODS file [{}].",
 					__func__,
 					byte_count,
 					self->irods_path_);
 			}
 
 			if (self->parser_->is_done()) {
-				if (self->upload_part_flag_) {
+				if (self->part_file_.is_open()) {
```
Just making sure... The only time `part_file_.is_open()` will be `true` is if `upload_part_flag_` is `true` as well, right?

JustinKyleJames replied:
Yes that is true.  This code also handles non-multipart uploads.

Also, even with `upload_part_flag_ = false`, the part file won't be open if we wrote that part directly to iRODS.

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
+			part_number_int = boost::lexical_cast<int>(part_number.c_str());
+		}
+		catch (const boost::bad_lexical_cast&) {
+			logging::error("{}: Upload ID {} Could not parse part_number ({})", __func__, upload_id, part_number);
```
```suggestion
			logging::error("{}: Upload ID [{}] Could not parse part_number [{}]", __func__, upload_id, part_number);
```

JustinKyleJames replied:
done

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
-			logging::error("{}: Upload ID {} was not in expected format.", __FUNCTION__, upload_id);
+			logging::error("{}: Upload ID {} was not in expected format.", __func__, upload_id);
+			response.result(beast::http::status::bad_request);
+			logging::debug("{}: returned {}", __func__, response.reason());
```
```suggestion
			logging::debug("{}: returned [{}]", __func__, response.reason());
```

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
+							if (old_part_size != part_size) {
+								// reject this
+								logging::error(
+									"{}: Upload ID {} - part_number {} was uploaded a second time with a different "
+									"part size. Old part size = {}. New part size = {}. "
```
```suggestion
									"{}: Upload ID [{}] - part_number [{}] was uploaded a second time with a different "
									"part size. Old part size = [{}]. New part size = [{}]. "
```

JustinKyleJames replied:
done

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
+							logging::error(
+								"{}: Upload ID {} - part_number {} was uploaded a second time with a different part "
+								"size. "
+								"Rejecting this request",
```
Agreed

JustinKyleJames replied:
done

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
+						part_offset += part_shmem::part_size_map[upload_id][i];
+					}
+					else {
+						know_part_offset = false;
```
If we get into this case, can we break out of the loop? Will we ever need the `part_offset` to be accurate if any of the previous `part_size`s are not known? Or am I misunderstanding?

JustinKyleJames replied:
We can break.  I made the change.

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
+				// Also note that the boost::beast::htpp parser will likely just reject this request so we may never get
+				// to the is point.
+				logging::warn(
+					"{}:{} {}: part_number = {} could not parse part size as unint64_t, "
```
```suggestion
					"{}:{} {}: part_number [{}] could not parse part size as unint64_t, "
```

JustinKyleJames replied:
done

alanking commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -402,33 +640,82 @@ void irods::s3::actions::handle_putobject(
-		if (upload_part) {
+		if (upload_part && know_part_offset) {
+			{
+				std::lock_guard<std::mutex> guard(part_shmem::part_size_map_mutex);
```
I don't see the `part_size_map` in here. Is this mutex also protecting the `replica_token_and_number_map`?

JustinKyleJames replied:
Yes it is.  I changed the mutex name to `multipart_global_state_mutex`.

alanking reviewed (commented):
I have some possibly dumb questions, but looks good overall.

JustinKyleJames commented:
> very impressive. glad this is behaving.
> 
> we have existing tests? new tests?

For the most part the existing tests should suffice.  I did have a problem with missing the last close which would cause a second put to that object to fail if done too quickly.  I just updated the tests to follow up the first large file put with a second one for each of the clients.

JustinKyleJames commented on endpoints/s3/src/abortmultipartupload.cpp:
```diff
@@ -0,0 +1,183 @@
+	logging::debug("{}: returned {}", __FUNCTION__, response.reason());
+	session_ptr->send(std::move(response));
+	return;
+}
```
done

JustinKyleJames commented on endpoints/s3/src/abortmultipartupload.cpp:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	if (upload_id.empty()) {
+		logging::error("{}: Did not receive a an uploadId", __func__);
```
done

JustinKyleJames commented on endpoints/s3/src/completemultipartupload.cpp:
```diff
@@ -92,44 +107,42 @@ void irods::s3::actions::handle_completemultipartupload(
 
 	if (upload_id.empty()) {
-		logging::error("{}: Did not receive a an uploadId", __FUNCTION__);
+		logging::error("{}: Did not receive a an uploadId", __func__);
```
done

JustinKyleJames commented on tests/docker/docker-compose.yml:
```diff
@@ -1,4 +1,5 @@
 version: '3'
```
done

JustinKyleJames commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -58,64 +98,135 @@ void manually_parse_chunked_body_write_to_irods_in_background(
+	~incremental_async_read()
+	{
+		if (keep_dstream_open_flag) {
+			//std::this_thread::sleep_for(std::chrono::seconds(3));
```
done

JustinKyleJames commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
+				// size in shared memory. This will mean all subsequent parts will have to be written to a local
+				// cache file and flushed on CompleteMultipartUpload which will impact performance.
+				//
+				// Also note that the boost::beast::htpp parser will likely just reject this request so we may never get
```
done

JustinKyleJames commented on endpoints/s3/src/putobject.cpp:
```diff
@@ -331,35 +446,158 @@ void irods::s3::actions::handle_putobject(
+				// cache file and flushed on CompleteMultipartUpload which will impact performance.
+				//
+				// Also note that the boost::beast::htpp parser will likely just reject this request so we may never get
+				// to the is point.
```
done

JustinKyleJames commented on core/src/session.cpp:
```diff
@@ -243,6 +243,17 @@ namespace irods::http
 					send(irods::http::fail(http::status::not_implemented));
 				}
+				else if (params.find("uploadId") != params.end()) {
+					logging::debug("{}: AbortMultipartUpload detected?", __func__);
```
This was just left over from when I was figuring it out.  I removed the question mark.

JustinKyleJames commented on endpoints/s3/src/abortmultipartupload.cpp:
```diff
@@ -0,0 +1,183 @@
+	// Do not allow an upload_id that is not in the format we have defined. People could do bad things
+	// if we didn't enforce this.
+	if (!std::regex_match(upload_id, upload_id_pattern)) {
+		logging::error("{}: Upload ID {} was not in expected format.", __func__, upload_id);
```
Done

JustinKyleJames commented on endpoints/s3/src/abortmultipartupload.cpp:
```diff
@@ -0,0 +1,183 @@
+			}
+		}
+	}
+	catch (std::exception&) {
+	}
```
I reorganized the code to log an error message and set the return code to internal_server_error.  Rather than the errors above I allow it to continue so that the shmem will be cleaned up.

I also separated the errors between filesystem iterator type exceptions and file removal exceptions so better error messages can be provided.

JustinKyleJames commented on endpoints/s3/src/abortmultipartupload.cpp:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	response.result(beast::http::status::ok);
+	logging::debug("{}: returned {}", __FUNCTION__, response.reason());
```
Old habits die hard.  Done.

JustinKyleJames commented on endpoints/s3/src/completemultipartupload.cpp:
```diff
@@ -92,44 +107,42 @@ void irods::s3::actions::handle_completemultipartupload(
 		return;
 	}
 
-	fs::client::create_collections(conn, path.parent_path());
```
This is a consequence of the multipart enhancement.  We now write parts directly to iRODS during part uploads, if possible.  Worst case scenario is where part 1 was uploaded last.  In this case it would be the only part written directly to iRODS.  So by the time we get here the parent collection would have already been created.

alanking replied:
Okay, makes sense. Resolving this.

JustinKyleJames commented on endpoints/s3/src/completemultipartupload.cpp:
```diff
@@ -192,21 +205,41 @@ void irods::s3::actions::handle_completemultipartupload(
 		return;
 	}
 
+	// debug
+	{
+		std::lock_guard<std::mutex> guard(part_shmem::part_size_map_mutex);
+
+		if (part_shmem::part_size_map.find(upload_id) != part_shmem::part_size_map.end()) {
+			logging::debug("{}:{} {}: ------------------", __FILE__, __LINE__, __func__);
+			for (int i = 1; i <= part_number_count; ++i) {
+				bool found = part_shmem::part_size_map[upload_id].find(i) != part_shmem::part_size_map[upload_id].end();
+				if (found) {
+					logging::debug(
+						"{}:{} {}: {}: {}", __FILE__, __LINE__, __func__, i, part_shmem::part_size_map[upload_id][i]);
+				}
+				else {
+					logging::debug("{}:{} {}: {}: UNKNOWN", __FILE__, __LINE__, __func__, i);
+				}
+			}
+			logging::debug("{}:{} {}: ------------------", __FILE__, __LINE__, __func__);
+		}
+	}
+
```
I put that in the following if block:

```
if (spdlog::get_level() == spdlog::level::debug || spdlog::get_level() == spdlog::level::trace)
```

Tested it and it works as expected.

JustinKyleJames commented on endpoints/s3/src/completemultipartupload.cpp:
```diff
@@ -404,27 +432,48 @@ void irods::s3::actions::handle_completemultipartupload(
+	if (part_shmem::replica_token_and_number_map.find(upload_id) != part_shmem::replica_token_and_number_map.end()) {
+		// Read all of the shared pointers in the tuple to make sure they are destructed in
+		// the order we require. std::tuple does not guarantee order of destruction.
+		auto conn_ptr = std::get<2>(part_shmem::replica_token_and_number_map[upload_id]);
+		auto transport_ptr = std::get<3>(part_shmem::replica_token_and_number_map[upload_id]);
```
I was troubleshooting an occasional error due to the order of deletion of these things when going through the chunk encoding flow.  This change was an attempt to fix it.  It turned out that I needed the following similar change in putobject.cpp:

```
            if (current_state == parsing_state::parsing_done) {                                                                                                                                                                                                                                                      
                // this is to force that these are destructed in the correct order                                                                                                                                                                                                                                   
                auto conn_ptr = conn;                                                                                                                                                                                                                                                                                
                auto transport_ptr = tp;                                                                                                                                                                                                                                                                             
                auto dstream_ptr = d;                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                     
                if (ofs->is_open()) {                                                                                                                                                                                                                                                                                
                    ofs->close();                                                                                                                                                                                                                                                                                    
                }                                                                                                                                                                                                                                                                                                    
                if (!keep_dstream_open_flag && d->is_open()) {                                                                                                                                                                                                                                                       
                    logging::trace("{}:{} Closing iRODS data object.", __func__, __LINE__);                                                                                                                                                                                                                          
                    d->close();                                                                                                                                                                                                                                                                                      
                }     
```
Once I made that change the occasional errors went away.  I left the similar code in completemultipartupload.cpp just to be safe.

The idea is that if I have a local copy of that pointer it won't be deleted when the `std::tuple` is destructed after `erase` and will instead be deleted in the reverse order of how these pointer copies are created when they go out of scope.

Why that is required when `close` has already been called is a mystery to me but it definitely seems to be required.  I am not familiar with the stream code to know exactly why this is the case.

Note that `std::tuple` does not state the order of destruction of its elements.  I tested this with clang and g++ and got difference results.  So I wanted to be able to control it thus I created local copies of these pointers to make sure these objects stay open and are deleted in the right order.

alanking replied:
Oh, I see. So having references to the shared pointers means that the shared pointers in the tuple are not the last references. Then the tuple gets destructed and these are the last remaining references, which means that they are destructed in reverse when leaving this scope (rather than leaving it up to "chance" in the tuple's destructing). That makes sense. Thanks

alanking commented on endpoints/s3/src/abortmultipartupload.cpp:
```diff
@@ -162,22 +165,33 @@ void irods::s3::actions::handle_abortmultipartupload(
 		}
 	}
 	catch (std::exception&) {
+		logging::error(
+			"{}: Upload ID [{}] - Exception caught when iterating through part files for removal.",
+			__func__,
+			upload_id);
```
May want to include the `what()` in this message

JustinKyleJames replied:
done

alanking reviewed (commented):
Sorry for the delay there. Fell through the cracks.

I noticed one thing but otherwise looks good to me

alanking reviewed (commented):
Yes, I think this is ready for squash

korydraughn commented:
Yes, proceed with squashing if it's passing the tests.

JustinKyleJames commented:
> Yes, proceed with squashing if it's passing the tests.

I have squashed.

Note that I also corrected my .vimrc file which was no longer highlighting spaces at the end of lines.  I noticed some of these in my docker-compose.yml so removed those.  Other than that this is identical to what you reviewed.

korydraughn commented:
Please capture the design/ideas behind the multipart enhancements in the commit body of 45508f5.

alanking reviewed (approved):
Please link new issues in a comment on this PR as @korydraughn said, and I think we're ready for #s. Maybe wait for one more approval

JustinKyleJames commented:
> Please capture the design/ideas behind the multipart enhancements in the commit body of [45508f5](https://github.com/irods/irods_client_s3_api/commit/45508f5c924b3795317867e9305b721bcdc45f11).

I added the design in the commit message.

korydraughn commented on README.md:
````diff
@@ -496,6 +496,8 @@ docker compose build
 docker compose run client
 ```
 
+> Note that if you get an error like "'name' does not match any of the regexes: '^x-'" then you will need to upgrade your version of docker compose.
````
Looked at the rendering of the README. In the section prior to this one, we include a note. Let's follow the same pattern as that one.

#### Pattern
- Make the statement italic
- Drop use of markdown quote (`>`)
- Replace `Note that` with `Note:`
- Highlight error message by wrapping in backticks

#### Result
_Note: If you get an error like `'name' does not match any of the regexes: '^x-'` then you will need to upgrade your version of docker compose._

JustinKyleJames replied:
I made the change and pounded.

korydraughn reviewed (approved):
Just one edit to the README.

Once handled, proceed with pounding.
